rir/src/compiler/analysis/scope.cpp

```diff
@@ -12,48 +12,72 @@ ScopeAnalysis::ScopeAnalysis(ClosureVersion* cls, Promise* prom, Value* promEnv,
     : StaticAnalysis("Scope", cls, prom, initialState, log), depth(depth),
       staticClosureEnv(promEnv) {}
 
-void ScopeAnalysis::lookup(const ScopeAnalysisState& state, Value* v,
-                           const LoadMaybe& action,
-                           const Maybe& notFound) const {
+void ScopeAnalysis::lookupGlobal(const ScopeAnalysisState& state, Value* v,
+                                 const LoadMaybe& action,
+                                 const Maybe& notFound) const {
     auto instr = Instruction::Cast(v);
     if (!instr)
         return notFound();
 
     // If the this is an call instruction or force we might have some result
     // value from the inter-procedural analysis.
+    // Since the "returnValues" are indexed by SSA variables, they are always
+    // valid, even if "state" does not correspond with the position of "v".
     if (state.returnValues.count(instr)) {
         auto& res = state.returnValues.at(instr);
+        if (res.isSingleValue()) {
+            lookupGlobal(state, res.singleValue().val, action,
+                         [&]() { action(AbstractLoad(res)); });
+            return;
+        }
         action(AbstractLoad(res));
         return;
     }
 
-    // IMPORTANT: Dead store elimination relies on the fact that we handle all
-    // possible loads here
+    notFound();
+}
 
-    // If this is a ldvar, we perform a get on the abstract environment and
-    // if possible recurse on the result
-    if (auto ld = LdVar::Cast(instr)) {
-        action(load(state, ld->varName, ld->env()));
-        return;
-    }
+void ScopeAnalysis::lookupAt(const ScopeAnalysisState& state, Value* v,
+                             const LoadMaybe& action,
+                             const Maybe& notFound) const {
 
-    // If this is a ldfun, we perform a getFun on the abstract environment and
-    // if possible recurse on the result. The loadFun is an abstract version of
-    // ldFun and considers the special case of skipping non-closure bindings.
-    // Thus it is a lot less reliable than normal load.
-    if (auto ldf = LdFun::Cast(instr)) {
-        action(loadFun(state, ldf->varName, ldf->env()));
-        return;
-    }
+    auto instr = Instruction::Cast(v);
+    if (!instr)
+        return notFound();
 
-    // If this is a ldvarsuper, we perform a superget on the abstract
-    // environment and if possible recurse on the result
-    if (auto sld = LdVarSuper::Cast(instr)) {
-        action(superLoad(state, sld->varName, sld->env()));
-        return;
-    }
+    lookupGlobal(state, instr, action, [&]() {
```

Comment:
Lookup calls lookupGlobal. So...I can't find the semantics difference between calling `lookupAt` and `lookupGlobal` in terms of the correspondence between the state and the current instruction